Automattic/wp-calypso
# Fix race conditions in localized SSR

Fixes a serious bug in localized server-side rendering discovered by @sgomes when testing the `/log-in` routes. It's caused by the fact that there is one global singleton instance of the `i18n` object on the server and all requests from all users share it.

1. Locally, run `yarn build-languages` so that your `public/languages/` folder is populated with language files like `id-v.1.1.json`.
2. Then run local Calypso with `yarn start`.
3. Disable JavaScript in your browser devtools so that you are sure that you always see only the server-rendered markup and that the client-side JavaScript doesn't run.
4. Open the localized `/id/themes` page. You should see Calypso localized in Indonesian.
5. Now open the `/log-in` page. As it doesn't have any locale suffix (like `/log-in/id`) it should be in English. But it's in Indonesian, the language of the previously rendered Themes page!

<img width="503" alt="Screenshot 2022-01-10 at 12 24 15" src="https://user-images.githubusercontent.com/664258/148758598-0bb716a5-d95a-476b-bb11-073eb7e943d7.png">

The reason is that the `ssrSetupLocale` page.js handler, used on the `/:lang/themes` page, will set the global `i18n` instance to Indonesian locale. Then, when the `/log-in` page is rendered, it doesn't modify the `i18n` instance and uses it, assuming it's using the default `en` locale`. But it doesn't. It uses the somewhat random locale that the last `/themes` request happened to request.

The fix has several parts:
- first, I had to clean up `i18n-calypso` TS types so that they declare the `I18N` class. TypeScript files wouldn't compile otherwise.
- then I added support for React context to `i18n-calypso`. The `localize` HOC and the `useTranslate` hook no longer use the global `i18n` instance directly, but instead use the one provided by `I18NContext`. The default global `i18n` instance is the default value of that context, so that unless you mount a provider, the library behaves exactly the same way as it did before.
- another thing I had to modify is the `/theme-filters` REST request issued by the `/themes` SSR. There is a `wpcom` handler that adds the `locale` query parameter to it, reading the locale value from the global `i18n` instance. That's going to always be `en` now. But we need to pass the real locale, because the filters response contains localized labels. I achieved this by adding a custom `locale` parameter to the `requestThemeFilters` action that overrides the `locale` param.
- finally I modified the `ssrSetupLocale` handler and the rendering code to create a local `i18n` instance and pass it in context.

**How to test:**
Verify that the bug described above is indeed fixed: the `/log-in` page is always English, no matter what request you did before.

Also verify that the `/id/themes` page continues to be localized:

<img width="1070" alt="Screenshot 2022-01-10 at 11 28 36" src="https://user-images.githubusercontent.com/664258/148759680-1c802320-43ae-48d5-baec-46a13fc08b35.png">

The labels on the toolbar (Majalah, Bisnis) come from the `/theme-filters?locale=id` server response, not from local `translate` calls, so verify that both sources of localization work correctly.
